### PR TITLE
Change the way we get environment variables

### PR DESCRIPTION
**Pull request by rw-access**
## Issues
Loosely related/discovered to this comment https://github.com/elastic/detection-rules/issues/17#issuecomment-690508065

## Summary
Got rid of click callback to use a non-interactive default variable. I think we were trying to do too much in a callback, and I believe any I/O in callback is discouraged.

### CODE REVIEW

File: detection_rules/eswrap.py [hunk 1]
```diff
@@ -12,7 +12,7 @@
 from kibana import Kibana, RuleResource
 
 from .main import root
-from .misc import set_param_values
+from .misc import environ
```
Comment (rw-access) [inline detection_rules/eswrap.py:15]: maybe getenv would be a better name?

Comment (rw-access) [reply detection_rules/eswrap.py]: eeb200d13e91e11fa544a25f697aec35f94f5c14

File: detection_rules/misc.py [hunk 2]
```diff
@@ -213,19 +213,6 @@ def parse_config():
...
-    if value:
-        return value
-    elif os.environ.get(env_key):
-        return os.environ[env_key]
-    elif config.get(key) is not None:
-        return config[key]
-    elif prompt:
-        return click.prompt(key, default=param.default if not param.default else None, hide_input=param.hide_input,
-                            show_default=True if param.default else False)
+def getenv(name):
```
Comment (brokensound77) [inline detection_rules/misc.py:216]: I like what this is intended to solve, but this totally eliminates the ability to use config files for anything.

**Comment by brokensound77** [comment]
We can remove all IO from the callback and then leave in the option to parse the value from a source based on precedence, which would still check for explicit, then envvar, then config option. Prompts can be moved to within their respective functions (as you did). 

Config parsing can be moved to the root group, where users can implicitly opt in to config parsing by storing in the default config file or optionally pass `-c` to explicitly opt in using a specific config file.

https://github.com/elastic/detection-rules/compare/main...brokensound77:tweak/var-setting?expand=1

Thoughts?

File: detection_rules/misc.py (hunk 2 above)
Comment (rw-access) [reply detection_rules/misc.py]: 7165756e5ba42e06aa8ed7b6bc48225cc9552ac2

File: detection_rules/eswrap.py [hunk 3]
```diff
@@ -229,16 +229,23 @@ def normalize_file(events_file):
...
+@click.option('--cloud-id', default=getdefault("cloud_id"))
+@click.option('--user', '-u', default=getdefault("user"))
+@click.option('--password', '-p', default=getdefault("password"))
 def kibana_upload(toml_files, kibana_url, cloud_id, user, password):
     """Upload a list of rule .toml files to Kibana."""
     from uuid import uuid4
     from .packaging import manage_versions
     from .schemas import downgrade
 
+    if not cloud_id or kibana_url:
```
Comment (brokensound77) [inline detection_rules/eswrap.py:242]: would need to add a similar blurb for `collect-events`

**Review by brokensound77 — APPROVED** [review]
One note, then LGTM

I would also verify that the [CLI.md](https://github.com/rw-access/detection-rules/blob/tweak/environ-getter/CLI.md#using-a-config-file-or-environment-variables) doesn't need any updating as a result of this

File: detection_rules/eswrap.py [hunk 4]
```diff
@@ -193,6 +193,13 @@ def collect_events(agent_hostname, elasticsearch_url, cloud_id, user, password,
     """Collect events from Elasticsearch."""
     match = {'agent.type': agent_type} if agent_type else {}
 
+    if not cloud_id or elasticsearch_url:
+        raise click.ClickException("Missing required --cloud-id or --elasticsearch-url")
+
+    # don't prompt for these until there's a cloud id or kibana URL
```
Comment (rw-access) [inline detection_rules/eswrap.py:199]:
```suggestion
    # don't prompt for these until there's a cloud id or elasticsearch URL
```

**Pull request merged**